File: scripts/compile-prompts.py
import os
import re
import json

PROMPTS_DIR = os.path.join(os.path.dirname(__file__), '../prompts')
OUTPUT_FILE = os.path.join(os.path.dirname(__file__), '../prompts.json')

def parse_markdown_file(filepath, filename):
    prompt_id = os.path.splitext(filename)[0]
    
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()
        
    # Generate default title from filename
    title = prompt_id.replace('-', ' ').replace('_', ' ').title()
    description = ""
    category = "General"
    tags = []
    prompt_body = ""
    
    # 1. Parse frontmatter if present
    frontmatter_match = re.match(r'^---\r?\n([\s\S]*?)\r?\n---\r?\n', content)
    
    if frontmatter_match:
        yaml_block = frontmatter_match.group(1)
        prompt_body = content[frontmatter_match.end():]
        
        for line in yaml_block.splitlines():
            if ':' in line:
                parts = line.split(':', 1)
                key = parts[0].strip().lower()
                val = parts[1].strip()
                
                # Strip wrapping quotes
                if (val.startswith('"') and val.endswith('"')) or (val.startswith("'") and val.endswith("'")):
                    val = val[1:-1]
                
                if key == 'title':
                    title = val
                elif key == 'description':
                    description = val
                elif key == 'category':
                    category = val
                elif key == 'tags':
                    tags = [t.strip() for t in val.split(',') if t.strip()]
    else:
        # No frontmatter: parse document structure
        lines = content.splitlines()
        title_index = -1
        desc_index = -1
        in_code_block = False
        code_block_lines = []
        fallback_lines = []
        
        for i, line in enumerate(lines):
            trimmed = line.strip()
            
            # Extract H1 or H2 heading
            if (trimmed.startswith('#') or trimmed.startswith('##')) and title_index == -1:
                title = re.sub(r'^#+\s*', '', trimmed)
                title_index = i
                continue
            
            # Extract description
            if title_index != -1 and len(trimmed) > 0 and desc_index == -1 and not trimmed.startswith('#') and not trimmed.startswith('```'):
                description = trimmed
                desc_index = i
                continue
            
            # Extract code blocks
            if trimmed.startswith('```'):
                in_code_block = not in_code_block
                continue
                
            if in_code_block:
                code_block_lines.append(line)
            elif title_index != -1 and i > title_index and i != desc_index:
                fallback_lines.append(line)
                
        if code_block_lines:
            prompt_body = '\n'.join(code_block_lines)
        else:
            prompt_body = '\n'.join(fallback_lines).strip()
            
    # 2. Extract first code block from body if wrapped in backticks
    code_block_match = re.search(r'```[a-zA-Z0-9_-]*\r?\n([\s\S]*?)\r?\n```', prompt_body)
    if code_block_match:
        prompt_body = code_block_match.group(1)
    else:
        prompt_body = prompt_body.strip()
        
    # 3. Extract variables with default values: {variable_name:default_value}
    # Variable name can be alphanumeric, dashes or underscores
    variable_regex = r'\{([a-zA-Z0-9_-]+)(?::([^}]+))?\}'
    variables_map = {}
    
    for match in re.finditer(variable_regex, prompt_body):
        var_name = match.group(1)
        default_val = match.group(2) if match.group(2) else ""
        
        # Register first encounter or update if default value is specified later
        if var_name not in variables_map or default_val != "":
            variables_map[var_name] = default_val
            
    variables = [{"name": name, "default": def_val} for name, def_val in variables_map.items()]
    
    return {
        "id": prompt_id,
        "title": title,
        "description": description,
        "category": category,
        "tags": tags,
        "prompt": prompt_body,
        "variables": variables
    }

def compile_prompts():
    print("Compiling prompts from Markdown files...")
    
    if not os.path.exists(PROMPTS_DIR):
        os.makedirs(PROMPTS_DIR)
        print(f"Created prompts directory: {PROMPTS_DIR}")
        
    compiled_list = []
    
    for filename in os.listdir(PROMPTS_DIR):
        if filename.endswith('.md'):
            filepath = os.path.join(PROMPTS_DIR, filename)
            try:
                parsed = parse_markdown_file(filepath, filename)
                compiled_list.append(parsed)
                print(f"- Parsed '{filename}' [Category: {parsed['category']}, Variables: {len(parsed['variables'])}]")
            except Exception as e:
                print(f"Error parsing '{filename}': {e}")
                
    with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
        json.dump(compiled_list, f, indent=2, ensure_ascii=False)
        
    print(f"\nCompilation complete! Saved {len(compiled_list)} prompts to: {OUTPUT_FILE}")

if __name__ == '__main__':
    compile_prompts()

File: scripts/compile-prompts.js
const fs = require('fs');
const path = require('path');

const PROMPTS_DIR = path.join(__dirname, '../prompts');
const OUTPUT_FILE = path.join(__dirname, '../prompts.json');

// Ensure output directories exist
if (!fs.existsSync(PROMPTS_DIR)) {
  fs.mkdirSync(PROMPTS_DIR, { recursive: true });
}

function parseMarkdownFile(filePath, filename) {
  const content = fs.readFileSync(filePath, 'utf8');
  const id = path.basename(filename, '.md');
  
  let title = id.replace(/[-_]/g, ' ');
  // Capitalize title words
  title = title.split(' ').map(w => w.charAt(0).toUpperCase() + w.slice(1)).join(' ');
  
  let description = '';
  let category = 'General';
  let tags = [];
  let promptBody = '';
  
  // 1. Check for frontmatter
  const frontmatterMatch = content.match(/^---\r?\n([\s\S]*?)\r?\n---\r?\n/);
  
  if (frontmatterMatch) {
    const yamlBlock = frontmatterMatch[1];
    promptBody = content.slice(frontmatterMatch[0].length);
    
    // Parse frontmatter line-by-line
    const lines = yamlBlock.split(/\r?\n/);
    lines.forEach(line => {
      const colonIndex = line.indexOf(':');
      if (colonIndex > -1) {
        const key = line.slice(0, colonIndex).trim().toLowerCase();
        let val = line.slice(colonIndex + 1).trim();
        
        // Strip wrapping quotes if present
        if ((val.startsWith('"') && val.endsWith('"')) || (val.startsWith("'") && val.endsWith("'"))) {
          val = val.slice(1, -1);
        }
        
        if (key === 'title') {
          title = val;
        } else if (key === 'description') {
          description = val;
        } else if (key === 'category') {
          category = val;
        } else if (key === 'tags') {
          tags = val.split(',').map(t => t.trim()).filter(t => t.length > 0);
        }
      }
    });
  } else {
    // No frontmatter: Parse text structure
    const lines = content.split(/\r?\n/);
    let titleIndex = -1;
    let descIndex = -1;
    let inCodeBlock = false;
    let codeBlockLines = [];
    let fallbackTextLines = [];
    
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i].trim();
      
      // Extract title from first H1/H2 heading
      if ((line.startsWith('#') || line.startsWith('##')) && titleIndex === -1) {
        title = line.replace(/^#+\s*/, '');
        titleIndex = i;
        continue;
      }
      
      // Parse description: first non-empty paragraph after title
      if (titleIndex !== -1 && line.length > 0 && descIndex === -1 && !line.startsWith('#') && !line.startsWith('```')) {
        description = line;
        descIndex = i;
        continue;
      }
      
      // Extract code block
      if (line.startsWith('```')) {
        inCodeBlock = !inCodeBlock;
        continue;
      }
      
      if (inCodeBlock) {
        codeBlockLines.push(lines[i]); // Keep original line indentation
      } else if (titleIndex !== -1 && i > titleIndex && i !== descIndex) {
        fallbackTextLines.push(lines[i]);
      }
    }
    
    if (codeBlockLines.length > 0) {
      promptBody = codeBlockLines.join('\n');
    } else {
      promptBody = fallbackTextLines.join('\n').trim();
    }
  }

  // 2. Extract first code block from promptBody if it's markdown-formatted promptBody
  const codeBlockMatch = promptBody.match(/```[a-zA-Z0-9_-]*\r?\n([\s\S]*?)\r?\n```/);
  if (codeBlockMatch) {
    promptBody = codeBlockMatch[1];
  } else {
    promptBody = promptBody.trim();
  }
  
  // 3. Extract variables with default values from the prompt body
  // Matches {variable} or {variable:default_value}
  const variableRegex = /\{([a-zA-Z0-9_-]+)(?::([^}]+))?\}/g;
  const variablesMap = new Map();
  let match;
  
  while ((match = variableRegex.exec(promptBody)) !== null) {
    const varName = match[1];
    const defaultValue = match[2] || '';
    
    // Store first occurrence or override with default if found later
    if (!variablesMap.has(varName) || defaultValue !== '') {
      variablesMap.set(varName, defaultValue);
    }
  }
  
  const variables = Array.from(variablesMap.entries()).map(([name, defVal]) => ({
    name,
    default: defVal
  }));

  return {
    id,
    title: title.trim(),
    description: description.trim(),
    category: category.trim(),
    tags,
    prompt: promptBody,
    variables
  };
}

function compile() {
  console.log('Compiling prompts...');
  try {
    const files = fs.readdirSync(PROMPTS_DIR);
    const compiledPrompts = [];
    
    files.forEach(file => {
      if (file.endsWith('.md')) {
        const filePath = path.join(PROMPTS_DIR, file);
        try {
          const parsed = parseMarkdownFile(filePath, file);
          compiledPrompts.push(parsed);
          console.log(`- Successfully parsed: ${file} [Category: ${parsed.category}, Variables: ${parsed.variables.length}]`);
        } catch (err) {
          console.error(`Error parsing ${file}:`, err);
        }
      }
    });
    
    fs.writeFileSync(OUTPUT_FILE, JSON.stringify(compiledPrompts, null, 2), 'utf8');
    console.log(`\nCompilation complete! Saved ${compiledPrompts.length} prompts to ${OUTPUT_FILE}`);
  } catch (err) {
    console.error('Failed to read prompts directory:', err);
    process.exit(1);
  }
}

compile();
